File: 11_动态规划/背包问题/4_分组背包/template/GroupKnapsack.ts
/* eslint-disable no-inner-declarations */

type Item = { value: number; weight: number }

/**
 * 分组背包，每组最多选一个.最大化价值.
 * @complexity O(n * m * maxCapacity)
 */
function groupKnapsackAtMostOne(groups: ArrayLike<ArrayLike<Item>>, maxCapacity: number): number {
  const dp = Array<number>(maxCapacity + 1).fill(0)
  const n = groups.length
  let maxJ = 0
  for (let i = 0; i < n; i++) {
    const group = groups[i]
    let curMax = 0
    for (let k = 0; k < group.length; k++) curMax = Math.max(curMax, group[k].weight)
    maxJ = Math.min(maxJ + curMax, maxCapacity)
    // 这里 j 的初始值可以优化成前 i 个组的每组最大重量之和（但不能超过 maxW）
    for (let j = maxJ; j >= 0; j--) {
      for (let k = 0; k < group.length; k++) {
        const item = group[k]
        if (item.weight <= j) {
          dp[j] = Math.max(dp[j], dp[j - item.weight] + item.value)
        }
      }
    }
  }

  return dp[maxCapacity]
}

/**
 * 分组背包，每组恰好选一个.最大化价值.
 * @returns dp[j] 表示从每组恰好选一个，能否凑成重量 j.
 * @complexity O(n * m * maxCapacity)
 */
function groupKnapsackExactOne(groups: ArrayLike<ArrayLike<number>>, maxCapacity: number): Uint8Array {
  const dp = new Uint8Array(maxCapacity + 1)
  dp[0] = 1
  const n = groups.length
  let maxJ = 0
  for (let i = 0; i < n; i++) {
    const group = groups[i]
    let curMax = 0
    for (let k = 0; k < group.length; k++) curMax = Math.max(curMax, group[k])
    maxJ = Math.min(maxJ + curMax, maxCapacity)
    // 这里 j 的初始值可以优化成前 i 个组的每组最大重量之和（但不能超过 maxW）
    for (let j = maxJ; j >= 0; j--) {
      let ok = false
      for (let k = 0; k < group.length; k++) {
        const weight = group[k]
        if (weight <= j && dp[j - weight]) {
          dp[j] = 1
          ok = true
          break
        }
      }
      if (!ok) dp[j] = 0
    }
  }
  return dp
}

export { groupKnapsackAtMostOne, groupKnapsackExactOne }

if (require.main === module) {
  // 1981. 最小化目标值与所选元素的差
  // https://leetcode.cn/problems/minimize-the-difference-between-target-and-chosen-elements/
  function minimizeTheDifference(mat: number[][], target: number): number {
    let maxSum = 0
    mat.forEach(row => {
      let max = 0
      for (let i = 0; i < row.length; i++) {
        max = Math.max(max, row[i])
      }
      maxSum += max
    })
    const ok = groupKnapsackExactOne(mat, maxSum)
    let res = 1e9
    for (let cand = 0; cand <= maxSum; cand++) {
      if (ok[cand]) {
        res = Math.min(res, Math.abs(cand - target))
      }
    }
    return res
  }

  // mat = [[1],[2],[3]],
  console.log(minimizeTheDifference([[1], [2], [3]], 100))
}

File: 11_动态规划/背包问题/4_分组背包/template/GroupKnapsack.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	分组背包问题()
}

// https://www.acwing.com/problem/content/9/
func 分组背包问题() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, maxCapacity int
	fmt.Fscan(in, &n, &maxCapacity)
	groups := make([][]item, n)
	for i := 0; i < n; i++ {
		var m int
		fmt.Fscan(in, &m)
		groups[i] = make([]item, m)
		for j := 0; j < m; j++ {
			var weight, value int
			fmt.Fscan(in, &weight, &value)
			groups[i][j] = item{value: value, weight: weight}
		}
	}

	res := GroupKnapsackAtMostOne(groups, maxCapacity)
	fmt.Fprintln(out, res)
}

type item struct{ value, weight int }

// 分组背包，每组最多选一个.最大化价值.
func GroupKnapsackAtMostOne(groups [][]item, maxCapacity int) int {
	dp := make([]int, maxCapacity+1)
	maxJ := 0
	for _, g := range groups {
		curMax := 0
		for _, w := range g {
			if w.weight > curMax {
				curMax = w.weight
			}
		}
		maxJ += curMax
		if maxJ > maxCapacity {
			maxJ = maxCapacity
		}
		// 这里 j 的初始值可以优化成前 i 个组的每组最大重量之和（但不能超过 maxW）
		for j := maxJ; j >= 0; j-- {
			for _, it := range g {
				if v, w := it.value, it.weight; w <= j {
					dp[j] = max(dp[j], dp[j-w]+v) // 如果 it.w 可能为 0 则需要用 dp[2][] 来滚动（或者保证每组至多一个 0 且 0 在该组最前面）
				}
			}
		}
	}
	return dp[maxCapacity]
}

// 分组背包，每组恰好选一个.最大化价值.
// 返回值：dp[j] 表示从每组恰好选一个，能否凑成重量 j.
// LC1981 https://leetcode-cn.com/problems/minimize-the-difference-between-target-and-chosen-elements/
func GroupKnapsackExactOne(groups [][]int, maxCapacity int) []bool {
	dp := make([]bool, maxCapacity+1) // dp[i][j] 表示能否从前 i 组物品中选出重量恰好为 j 的，且每组都恰好选一个物品
	dp[0] = true
	maxJ := 0
	for _, g := range groups {
		curMax := 0
		for _, w := range g {
			if w > curMax {
				curMax = w
			}
		}
		maxJ += curMax
		if maxJ > maxCapacity {
			maxJ = maxCapacity
		}
		// 这里 j 的初始值可以优化成前 i 个组的每组最大重量之和（但不能超过 maxW）
		for j := maxJ; j >= 0; j-- {
			ok := false
			for _, w := range g {
				if w <= j && dp[j-w] {
					dp[j] = true
					ok = true
					break
				}
			}
			if !ok {
				dp[j] = false
			}
		}
	}
	return dp
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// https://leetcode-cn.com/problems/minimize-the-difference-between-target-and-chosen-elements/
func minimizeTheDifference(mat [][]int, target int) int {
	maxSum := 0
	for _, row := range mat {
		max := 0
		for _, v := range row {
			if v > max {
				max = v
			}
		}
		maxSum += max
	}

	ok := GroupKnapsackExactOne(mat, maxSum)
	res := math.MaxInt
	for cand := range ok {
		if ok[cand] {
			res = min(res, abs(cand-target))
		}
	}
	return res
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
